runtime/src/bank.rs

```diff
@@ -15039,114 +15039,162 @@ pub(crate) mod tests {
+            assert_eq!(bank.get_account_modified_slot(&program_id).unwrap().1, slot);
 
-        let bank = Arc::new(new_from_parent(&bank));
-        // When replacing builtin_program, name must change to disambiguate from repeated
-        // invocations.
-        assert_capitalization_diff(
-            &bank,
-            || bank.add_builtin_account("mock_program v2", &program_id, true),
-            |old, new| {
-                assert_eq!(old, new);
-                true
-            },
-        );
+            let bank = Arc::new(new_from_parent(&bank));
+            add_root_and_flush_write_cache(&bank.parent().unwrap());
+            // When replacing builtin_program, name must change to disambiguate from repeated
+            // invocations.
+            assert_capitalization_diff(
+                &bank,
+                || bank.add_builtin_account("mock_program v2", &program_id, true),
+                |old, new| {
+                    assert_eq!(old, new);
+                    pass == 2
+                },
+            );
+            if pass == 2 {
+                continue;
+            }
 
-        assert_eq!(
-            bank.get_account_modified_slot(&program_id).unwrap().1,
-            bank.slot()
-        );
+            assert_eq!(
+                bank.get_account_modified_slot(&program_id).unwrap().1,
+                bank.slot()
+            );
 
-        let bank = Arc::new(new_from_parent(&bank));
-        assert_capitalization_diff(
-            &bank,
-            || bank.add_builtin_account("mock_program v2", &program_id, true),
-            |old, new| {
-                assert_eq!(old, new);
-                true
-            },
-        );
+            let bank = Arc::new(new_from_parent(&bank));
+            add_root_and_flush_write_cache(&bank.parent().unwrap());
+            assert_capitalization_diff(
+                &bank,
+                || bank.add_builtin_account("mock_program v2", &program_id, true),
+                |old, new| {
+                    assert_eq!(old, new);
+                    pass == 3
+                },
+            );
+            if pass == 3 {
+                continue;
+            }
 
-        // replacing with same name shouldn't update account
-        assert_eq!(
-            bank.get_account_modified_slot(&program_id).unwrap().1,
-            bank.parent_slot()
-        );
+            // replacing with same name shouldn't update account
+            assert_eq!(
+                bank.get_account_modified_slot(&program_id).unwrap().1,
+                bank.parent_slot()
+            );
+        }
     }
 
+    /// useful to adapt tests written prior to introduction of the write cache
+    /// to use the write cache
     fn add_root_and_flush_write_cache(bank: &Bank) {
-        bank.rc.accounts.add_root(bank.slot());
-        bank.flush_accounts_cache_slot_for_tests()
+        if bank.rc.accounts.accounts_db.caching_enabled {
```

Comment:
Idea is that it is cumbersome to change individual tests to use the write cache and then cumbersome to remove that special code later when using the write cache is the only option. The tests are running now and passing and testing the non-cache code path. I'd like to make the switch for all tests to use the write cache. But, some tests are failing. Each test fits into one of a few common methods to fix it. Adding calls to `add_root_and_flush_write_cache` is usually part of the fix. But, this call can be made innocuous and even skipped if it is called and we're not caching anyway (because a test isn't updated). So, I can test with the global flag turned on to make ALL tests run with the write cache on. I can then fix individual tests so they are flexible to run with or without the write cache. Then, when the global flag is switched, no code needs to be changed later. Most tests already run either way. I just have to fix up some tests which will need modification (like these) when I turn the write cache on by default for tests.